### PR TITLE
AJ-1092:  Show PFB Schema

### PR DESCRIPTION
**Pull request by snf2ye**

### CODE REVIEW

File: library/src/main/java/bio/terra/pfb/PfbReader.java [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+  }
+
+  Schema readFilePathPFBSchema(String fileLocation) throws IOException {
+    DatumReader<Entity> datumReader = new SpecificDatumReader<>(Entity.class);
+    DataFileReader<Entity> dataFileReader =
+        new DataFileReader<>(new File(fileLocation), datumReader);
+    return dataFileReader.getSchema();
+  }
+
+  public Metadata deserializeAvroFile(String fileLocation) throws IOException {
```
Comment (calypsomatic) [inline library/src/main/java/bio/terra/pfb/PfbReader.java:60]: Is an avro file just Metadata?  Would it make sense to name this something like 'getMetadata' since that's what it's aways doing, or is there never a case where you would deserialize and avro file and get something other than metadata?

File: library/src/main/java/bio/terra/pfb/PfbReader.java [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+    // One of these entities must be a "Metadata" object (Also defined in sample.advl)
+    // The rest of the entities are the data entries
+    // TODO: Is it possible to get in an infinite loop here if there is no metadata object?
+    while (dataFileReader.hasNext()) {
+      // One of the elements must be a "Metadata" object. From this, we can get the schema for the
+      // data
+      try {
+        data = dataFileReader.next(data);
+        result = (Metadata) data.getObject();
+        if (result != null) {
```
Comment (calypsomatic) [inline library/src/main/java/bio/terra/pfb/PfbReader.java:77]: Just expanding on the TODO here - I ran through the test and it looked like every one of the test files got a result from the first item from the dataFileReader iteration - so could be good to have test files that don't have a metadata object at all as well as not the first item.

File: library/src/test/java/bio/terra/pfb/PfbReaderTest.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+import java.io.IOException;
+import java.util.List;
+import org.junit.jupiter.api.Disabled;
+import org.junit.jupiter.api.Test;
+
+class PfbReaderTest {
+
+  @Test
+  void showSchemaTest() throws IOException {
+    var listOfTestFiles = List.of("minimal_data", "minimal_schema", "test", "kf");
```
Comment (calypsomatic) [inline library/src/test/java/bio/terra/pfb/PfbReaderTest.java:13]: Bit of a nit: looks like all the test files are the same between tests, could we pull them out into a class variable?

File: library/src/main/java/bio/terra/pfb/PfbReader.java (hunk 1 above)
Comment (snf2ye) [reply library/src/main/java/bio/terra/pfb/PfbReader.java:60]: This method focuses on Metadata, so yes, good idea to rename it. 

PFB files are Avro files that follow a specific schema. The schema includes a little of "Entity" objects, where an entity can either be a "Metadata" object or a table row object.  I've tried to detail this setup in this diagram and there are more details in the [pyPFB library](https://github.com/uc-cdis/pypfb/blob/master/docs/index.md#introduction). 
PFB

File: library/src/main/java/bio/terra/pfb/PfbReader.java (hunk 2 above)
Comment (snf2ye) [reply library/src/main/java/bio/terra/pfb/PfbReader.java:77]: After reading more about this, I think we can just throw an exception if the first entry is not a metadata entity. [The spec requires the first entity meets the specs for a Metadata entity](https://github.com/uc-cdis/pypfb/blob/master/docs/index.md#schema).

File: library/src/test/java/bio/terra/pfb/PfbReaderTest.java (hunk 3 above)
Comment (snf2ye) [reply library/src/test/java/bio/terra/pfb/PfbReaderTest.java:13]: Yes, can do!

File: library/src/test/java/bio/terra/pfb/utils/PfbJsonComparator.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    }
+    if ((o1 instanceof TextNode) && (o2 instanceof TextNode)) {
+      String s1 = ((TextNode) o1).asText();
+      String s2 = ((TextNode) o2).asText();
+
+      if (s1.equalsIgnoreCase(s2)) {
+        return 0;
+      }
+    }
+    return 1;
```
[the remark this answers is not in the document]
Comment (snf2ye) [reply library/src/test/java/bio/terra/pfb/utils/PfbJsonComparator.java:21]: Ok, I've actually removed this -- I was using it for debugging to try to pinpoint where pyPFB output differs from the java-pfb output.

File: cli/README.md [hunk 5]
```diff
@@ -2,11 +2,43 @@ Java PFB CLI
 ===
 
 ## Available Commands
-- hello
-- getNumber5
+- show
 - --version
 - --help
 
+Examples:
```
Comment (jladieu) [inline cli/README.md:9]: Great usage documentation!

File: cli/src/main/java/bio/terra/pfb/JavaPfbCommand.java [hunk 6]
```diff
@@ -10,6 +12,29 @@
...
 public class JavaPfbCommand implements Runnable {
+
+  @Parameters(index = "0", description = "command to run (Options include: show)")
+  private PfbCommand command;
+
+  @Parameters(
+      index = "1",
+      description =
+          "optional option to run for given command (Options include: schema, metadata, nodes)",
+      defaultValue = "tableRows")
```
Comment (jladieu) [inline cli/src/main/java/bio/terra/pfb/JavaPfbCommand.java:23]: Is `tableRows` supposed to the default here?  I was expecting one of `schema`, `metadata`, or `nodes`?

File: cli/src/main/java/bio/terra/pfb/JavaPfbCommand.java [hunk 7]
```diff
@@ -21,19 +46,70 @@ static int executeCommand(String[] args) {
 
   @Override
   public void run() {
+    Library library = new Library(new PfbReader());
     System.out.println("PFB RUN");
+    if (command != null && option != null) {
+      if (command.equals(SHOW)) {
+        switch (option) {
+          case SCHEMA:
+            System.out.println("Show schema for file path: " + filePath);
```
Comment (jladieu) [inline cli/src/main/java/bio/terra/pfb/JavaPfbCommand.java:55]: No need to change here, but it might be worth considering extracting all logging into a logger seam so that we could override the behavior to be something other than `System.out`.  Does this seem reasonable to you as a possible future change, or is there something that makes `System.out` a requirement?

File: cli/src/test/java/bio/terra/pfb/JavaPfbCommandTest.java [hunk 8]
```diff
@@ -37,42 +38,51 @@ void run() {
...
 
   @Test
   void testCorrectVersion() {
     GitConfiguration gitConfiguration = new GitConfiguration();
-    String exepectedVersion = gitConfiguration.getCliVersion();
-    String[] args = new String[1];
-    args[0] = "--version";
-    javaPfbCommand.executeCommand(args);
-    assertThat(outContent.toString(), equalTo("pfb " + exepectedVersion + "\n"));
+    String expectedVersion = gitConfiguration.getCliVersion();
```
Comment (jladieu) [inline cli/src/test/java/bio/terra/pfb/JavaPfbCommandTest.java:60]: Love the level of abstraction for the test cases here.  Great work!

File: library/src/main/java/bio/terra/pfb/PfbReader.java [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+    try {
+      Schema schema = getSchema(fileLocation);
+      List<Schema> pfbSchemaShow =
+          schema.getField("object").schema().getTypes().stream()
+              .filter(t -> !t.getName().equals("Metadata"))
+              .toList();
+      String show = pfbSchemaShow.stream().map(Schema::toString).toList().toString();
+      return convertEnum(show);
+    } catch (IOException e) {
+      return getErrorMessage(e);
```
Comment (jladieu) [inline library/src/main/java/bio/terra/pfb/PfbReader.java:34]: No change needed here, but is it important for this library _not_ to throw exceptions?  My instinct would be to let these exceptions bubble up assuming they are truly abnormal behavior, and let the CLI layer decide whether or not to convert them to text output.  Wondering if you think this would be a reasonable change.

File: library/src/main/java/bio/terra/pfb/PfbReader.java [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+    (https://avro.apache.org/docs/1.9.1/spec.html#Enums), all enums are encoded in
+    such a way, that all other symbols is being encoded with codepoint wrapped in
+    single underscores. For example bpm > 60 will be encoded in: bpm_32__62__32_60,
+    so space   is encoded as _32_ and greater sign > into _62_. Same for Unicode
+    characters: ä - _228_, ü - _252_. The Avro schema also doesn't allow for the
+    first character to be a number. So we encode the first character in the way if
+    the character happens to be a number."
+  */
+  private String convertEnum(String schema) {
+    HashMap<String, String> map = new HashMap<>();
```
Comment (jladieu) [inline library/src/main/java/bio/terra/pfb/PfbReader.java:160]: Is there a reference / source of truth somewhere that we can check going forward to make sure this map is comprehensive?  I imagine it could become outdated if additional characters/encodings were introduced.

File: library/src/test/java/bio/terra/pfb/PfbReaderTest.java [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+
+class PfbReaderTest {
+
+  private static final List<String> listOfTestFiles =
+      List.of("minimal_schema", "test", "kf", "data-with-array", "empty");
+
+  @Test
+  void showSchemaTest() throws IOException {
+    // TODO - remove when fix is added for AJ-1288
+    var editedListOfTestFiles = listOfTestFiles.stream().filter(f -> !f.equals("empty")).toList();
```
Comment (jladieu) [inline library/src/test/java/bio/terra/pfb/PfbReaderTest.java:21]: What's the significance of the `empty` file here, and why does it need to be excluded?

File: library/src/test/java/bio/terra/pfb/utils/PfbJsonComparator.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (jladieu) [reply library/src/test/java/bio/terra/pfb/utils/PfbJsonComparator.java:25]: Good idea!  I think renaming it to something that doesn't imply conformity to `Comparator` interface makes perfect sense, eg: `compareIgnoringOrdering`; this will distinguish it from the more rigid requirements of `Comparator`.

**Review by jladieu — APPROVED** [review]
I love the test coverage and the code is clean and should be easy to pick up.

I only had a couple spots where my pattern matching detected something I wanted to ask about. 

Otherwise, there are some exploratory questions around potential downstream changes but I've hopefully clearly marked those as not needing any changes :)

File: cli/src/main/java/bio/terra/pfb/JavaPfbCommand.java (hunk 6 above)
Comment (snf2ye) [reply cli/src/main/java/bio/terra/pfb/JavaPfbCommand.java:23]: Thanks for calling this out -- I wasn't quite sure how to handle this case.  I'm trying to follow the same command syntax as used in pyPFB. In pyPFB, when there isn't a second argument for the "show" command, it defaults to display the table row data. So, I want this field to be optional and when it's not defined to be set to tableRows.

File: library/src/main/java/bio/terra/pfb/PfbReader.java (hunk 10 above)
Comment (snf2ye) [reply library/src/main/java/bio/terra/pfb/PfbReader.java:160]: Definitely not the best solution! This code is a holdover until addressed the root issue is addressed in AJ-1288.

File: library/src/test/java/bio/terra/pfb/PfbReaderTest.java (hunk 11 above)
Comment (snf2ye) [reply library/src/test/java/bio/terra/pfb/PfbReaderTest.java:21]: Definitely confusing -- "Empty" is the name of a test file that is definitely not empty. I'll rename it. 
But, this showSchema test fails for the test file "empty" and the root issue will be address in AJ-1288.

File: library/src/main/avro/pfbSchema.avdl [hunk 12]
```diff
@@ -0,0 +1,55 @@
+@namespace("bio.terra.pfb")
+/**
+ *  Built based on example provided by the pyPFB library
+ https://github.com/uc-cdis/pypfb/blob/master/docs/index.md#example
+ For PFB files, the only entity which is MUST be included in the PFB file is Metadata: this
```
Comment (calypsomatic) [inline library/src/main/avro/pfbSchema.avdl:5]: Xtreme nit: "which is MUST" has an extra "is"!  Only update if you're already making other changes!

File: library/src/test/java/bio/terra/pfb/LibraryTest.java [hunk 13]
```diff
@@ -1,14 +1,85 @@
...
+  private static final String FAKE_DATA_RETURN = "Fake data return";
+  @Mock PfbReader pfbReader;
 
   @Test
-  void testStubMethod() {
-    assertThat(Library.getNumber5(), equalTo(5));
+  void showSchemaTest() throws IOException {
+    when(pfbReader.showSchema(TEST_FILE_LOCATION)).thenReturn(FAKE_DATA_RETURN);
+    Library library = new Library(pfbReader);
+    assertReturnValueMatches(library.showSchema(TEST_FILE_LOCATION));
```
Comment (calypsomatic) [inline library/src/test/java/bio/terra/pfb/LibraryTest.java:28]: These are just verifying that the library shows the same info as the pfbreader?

File: cli/src/main/java/bio/terra/pfb/JavaPfbCommand.java (hunk 6 above)
Comment (snf2ye) [reply cli/src/main/java/bio/terra/pfb/JavaPfbCommand.java:23]: Good point --added!

File: library/src/main/avro/pfbSchema.avdl (hunk 12 above)
Comment (snf2ye) [reply library/src/main/avro/pfbSchema.avdl:5]: Thanks! I've updated it!

File: library/src/test/java/bio/terra/pfb/LibraryTest.java (hunk 13 above)
Comment (snf2ye) [reply library/src/test/java/bio/terra/pfb/LibraryTest.java:28]: Yeah - I suppose I'm confirming that the library methods are wired up to the correct pfbreader methods and that the data/file location is correctly getting passed through. 
The sonarcloud code coverage metric might be gamifying this to much... haha Mostly kidding -- Good to have sanity checks in place that things are wired up correctly.

File: library/src/test/java/bio/terra/pfb/utils/PfbJsonComparator.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (snf2ye) [reply library/src/test/java/bio/terra/pfb/utils/PfbJsonComparator.java:25]: Sounds good!

File: cli/src/main/java/bio/terra/pfb/JavaPfbCommand.java [hunk 14]
```diff
@@ -21,19 +52,69 @@ static int executeCommand(String[] args) {
...
   @Override
   public void run() {
-    System.out.println("PFB RUN");
+    Library library = new Library(new PfbReader());
+    logger.info("PFB RUN");
+    if (command != null && option != null) {
+      if (command.equals(SHOW)) {
+        switch (option) {
+          case SCHEMA:
+            logger.info("Show schema for file path: {}", filePath);
```
Comment (pshapiro4broad) [inline cli/src/main/java/bio/terra/pfb/JavaPfbCommand.java:61]: It's up to you, but I don't think `logger` is the correct tool for producing output from a command line tool, unless the you want the output to look like a log file.

File: library/src/test/java/bio/terra/pfb/utils/CompareOutputUtils.java [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+      assertTrue(pyPfbJsonOutput.equals(cmp, showSchemaOutputJson));
+      //      assertThat(pyPfbJsonOutput, equalTo(showSchemaOutputJson));
+    } else if (fileExtension.equals(TXT)) {
+      assertThat(pythonOutput, equalTo(javaPfbOutput));
+    } else {
+      throw new InvalidObjectException("Invalid file extension");
+    }
+  }
+
+  public static void compareJSONLineByLine(
```
Comment (pshapiro4broad) [inline library/src/test/java/bio/terra/pfb/utils/CompareOutputUtils.java:64]: There are libraries for writing tests that compare JSON, it might be better to use them here rather than writing your own. The one I've used is [JSONassert](https://github.com/skyscreamer/JSONassert)

**Review by pshapiro4broad — COMMENTED** [review]
I have a few comments, mainly minor things.

File: cli/src/main/java/bio/terra/pfb/JavaPfbCommand.java (hunk 7 above)
Comment (snf2ye) [reply cli/src/main/java/bio/terra/pfb/JavaPfbCommand.java]: That is a great point -- For some reason I got in my head that I couldn't use a logger because I needed the standard output for the CLI. But, I've added a logback integration that outputs to the console. Thanks!

File: cli/src/main/java/bio/terra/pfb/JavaPfbCommand.java [hunk 16]
```diff
@@ -10,6 +12,29 @@
...
+
+  @Parameters(index = "0", description = "command to run (Options include: show)")
+  private PfbCommand command;
+
+  @Parameters(
+      index = "1",
+      description =
+          "optional option to run for given command (Options include: schema, metadata, nodes)",
+      defaultValue = "tableRows")
+  private PfbCommandOption option = TABLE_ROWS;
```
[the remark this answers is not in the document]
Comment (snf2ye) [reply cli/src/main/java/bio/terra/pfb/JavaPfbCommand.java:30]: Oh you're right, it should only be done one way.

File: cli/src/main/java/bio/terra/pfb/JavaPfbCommand.java [hunk 17]
```diff
@@ -21,19 +46,70 @@ static int executeCommand(String[] args) {
 
   @Override
   public void run() {
+    Library library = new Library(new PfbReader());
     System.out.println("PFB RUN");
+    if (command != null && option != null) {
```
[the remark this answers is not in the document]
Comment (snf2ye) [reply cli/src/main/java/bio/terra/pfb/JavaPfbCommand.java]: Yes, looks like the ["arity"](https://picocli.info/#_arity) parameter can make it required. 
With these changes, usage complains about the missing parameter:
image

File: cli/src/main/java/bio/terra/pfb/JavaPfbCommand.java [hunk 18]
```diff
@@ -21,19 +46,70 @@ static int executeCommand(String[] args) {
...
+      if (command.equals(SHOW)) {
+        switch (option) {
+          case SCHEMA:
+            System.out.println("Show schema for file path: " + filePath);
+            System.out.println(library.showSchema(filePath));
+            break;
+          case TABLE_ROWS:
+            if (limit >= 0) {
+              System.out.println(
+                  "show table rows for file path: " + filePath + ", Limit = " + limit);
```
[the remark this answers is not in the document]
Comment (snf2ye) [reply cli/src/main/java/bio/terra/pfb/JavaPfbCommand.java]: This changed with the switch to use a logger.

File: library/src/main/java/bio/terra/pfb/Library.java [hunk 19]
```diff
@@ -1,12 +1,31 @@
 package bio.terra.pfb;
 
+import java.util.stream.Collectors;
+
 public class Library {
+  private PfbReader pfbReader;
```
[the remark this answers is not in the document]
Comment (snf2ye) [reply library/src/main/java/bio/terra/pfb/Library.java:6]: I've converted it to a record!

File: library/src/main/java/bio/terra/pfb/PfbReader.java [hunk 20]
```diff
@@ -0,0 +1,211 @@
...
+    map.put("_5f_", "_");
+    map.put("_60_", "`");
+    map.put("_7b_", "{");
+    map.put("_7c_", "|");
+    map.put("_7d_", "}");
+    map.put("_7e_", "~");
+
+    for (Map.Entry<String, String> entry : map.entrySet()) {
+      if (schema.contains(entry.getKey())) {
+        schema = schema.replace(entry.getKey(), entry.getValue());
```
[the remark this answers is not in the document]
Comment (snf2ye) [reply library/src/main/java/bio/terra/pfb/PfbReader.java:168]: True - I'll switch to automatically replace without checking contains.

File: library/src/main/java/bio/terra/pfb/PfbReader.java [hunk 21]
```diff
@@ -0,0 +1,211 @@
...
+      return getErrorMessage(e);
+    }
+  }
+
+  public Schema getSchema(String fileLocation) throws IOException {
+    boolean isUrl = isValidUrl(fileLocation);
+    if (isUrl) {
+      return readUrlPFBSchema(fileLocation);
+    }
+    return readFilePathPFBSchema(fileLocation);
```
[the remark this answers is not in the document]
Comment (snf2ye) [reply library/src/main/java/bio/terra/pfb/PfbReader.java:49]: True - that works! Thanks.

File: library/src/main/java/bio/terra/pfb/PfbReader.java [hunk 22]
```diff
@@ -0,0 +1,211 @@
...
+      return List.of(getErrorMessage(e));
+    }
+  }
+
+  public Metadata getPFBMetadata(String fileLocation) throws IOException {
+    File pfbData = new File(fileLocation);
+    // Deserialize the above generated avro data file
+    DatumReader<Entity> datumReader = new SpecificDatumReader<>(Entity.class);
+    Entity data = null;
+    Metadata result = null;
```
[the remark this answers is not in the document]
Comment (snf2ye) [reply library/src/main/java/bio/terra/pfb/PfbReader.java:105]: Oh, good point. I think at in a previous iteration I needed the variable defined outside of the try block, but not anymore!

File: library/src/main/java/bio/terra/pfb/PfbReader.java [hunk 23]
```diff
@@ -0,0 +1,211 @@
...
+    }
+    throw new InvalidPfbException("Error reading PFB Metadata object");
+  }
+
+  Schema readUrlPFBSchema(String signedUrl) throws IOException {
+    Schema schema;
+    DatumReader<Entity> datumReader = new SpecificDatumReader<>(Entity.class);
+    try (InputStream in = readFromSignedUrl(signedUrl);
+        DataFileStream<Entity> reader = new DataFileStream<>(in, datumReader)) {
+      schema = reader.getSchema();
```
[the remark this answers is not in the document]
Comment (snf2ye) [reply library/src/main/java/bio/terra/pfb/PfbReader.java:125]: True - much cleaner!

File: library/src/main/java/bio/terra/pfb/PfbReader.java [hunk 24]
```diff
@@ -0,0 +1,211 @@
...
+  public String showMetadata(String fileLocation) {
+    try {
+      Metadata metadata = getPFBMetadata(fileLocation);
+      return metadata.toString();
+    } catch (Exception e) {
+      return getErrorMessage(e);
+    }
+  }
+
+  Schema readFilePathPFBSchema(String fileLocation) throws IOException {
```
[the remark this answers is not in the document]
Comment (snf2ye) [reply library/src/main/java/bio/terra/pfb/PfbReader.java:71]: Ah good catch - I thought I capitalized them all correctly, but definitely missed a few. Thanks!

File: library/src/main/java/bio/terra/pfb/Library.java [hunk 25]
```diff
@@ -1,12 +1,31 @@
...
-  private Library() {
-    // intentionally empty
+  public String showTableRows(String fileLocation) {
+    return pfbReader.show(fileLocation).stream().collect(Collectors.joining("\n"));
   }
 
-  public static int getNumber5() {
-    return 5 * 1;
+  public String showTableRows(String fileLocation, int limit) {
+    return pfbReader.show(fileLocation).stream().limit(limit).collect(Collectors.joining("\n"));
```
[the remark this answers is not in the document]
Comment (snf2ye) [reply library/src/main/java/bio/terra/pfb/Library.java:24]: Once you add in the complexity of dealing with the edge case that limit is greater than the length of the list, the original code seems more clear to me.

File: library/src/test/java/bio/terra/pfb/LibraryTest.java [hunk 26]
```diff
@@ -1,14 +1,85 @@
...
+
+    String oneElementLimitedShow = library.showTableRows(TEST_FILE_LOCATION, 1);
+    testCorrectNumberReturned(oneElementLimitedShow, 1);
+    String threeElementLimitedShow = library.showTableRows(TEST_FILE_LOCATION, 3);
+    testCorrectNumberReturned(threeElementLimitedShow, 3);
+  }
+
+  private void testCorrectNumberReturned(String returnedVal, int expectedNum) {
+    List<String> elements = Arrays.stream(returnedVal.split("\n")).toList();
+    assertThat("Correct number of elements returned", elements.size(), equalTo(expectedNum));
```
[the remark this answers is not in the document]
Comment (snf2ye) [reply library/src/test/java/bio/terra/pfb/LibraryTest.java:68]: Ah ok, great!

File: library/src/test/java/bio/terra/pfb/utils/CompareOutputUtils.java [hunk 27]
```diff
@@ -0,0 +1,139 @@
...
+        break;
+      case SHOW_METADATA:
+        javaPfbOutput = reader.showMetadata(avroFilePath);
+        break;
+      case SHOW_NODES:
+        javaPfbOutput = reader.showNodes(avroFilePath);
+        break;
+      default:
+        throw new InvalidObjectException("Invalid command type");
+    }
```
[the remark this answers is not in the document]
Comment (snf2ye) [reply library/src/test/java/bio/terra/pfb/utils/CompareOutputUtils.java:49]: Oh, that's much better.

File: library/src/test/java/bio/terra/pfb/utils/CompareOutputUtils.java (hunk 15 above)
Comment (snf2ye) [reply library/src/test/java/bio/terra/pfb/utils/CompareOutputUtils.java:64]: I'm able to more directly use the Object Mapper to compare JSON for most cases. This is a special case because each line in the file is a json object, rather than an entire json file.

File: cli/src/main/java/bio/terra/pfb/JavaPfbCommand.java (hunk 14 above)
Comment (snf2ye) [reply cli/src/main/java/bio/terra/pfb/JavaPfbCommand.java]: I'm a little torn here. 
I do like using the logback appender since it's a standard across our java projects and it has helpful error message formatting built-in.
It does seem like we can format the logs however desired with the logback yaml. 

But, PrintStream or System.out should also both be pretty straightforward.

Comment (pshapiro4broad) [reply cli/src/main/java/bio/terra/pfb/JavaPfbCommand.java]: I think either could work. IMO the logging API's output isn't good CLI output in that it has timestamps and other things that a usual CLI user wouldn't be interested in (which can be customized) . It also requires configuration and is overall less flexible than using a `PrintStream`. In a spring boot app `System.out` use is discouraged but I think in a CLI app it makes sense.

It's true that the log APIs make it easy to include a stack trace in the output, although I think for a CLI a "exceptional" exception shouldn't be caught by the app code. The top-level exception handler will catch it an print the error. In the case where an exception is expected I think you'd want to catch it and print out a custom message anyway.

On the other hand, I've seen long running CLIs that are multithreaded which use a logger for their output. If the CLI is running long enough or needs lots of diagnostic output then it could be useful.

File: library/src/main/java/bio/terra/pfb/PfbReader.java (hunk 9 above)
Comment (pshapiro4broad) [reply library/src/main/java/bio/terra/pfb/PfbReader.java:39]: It looks like picocli also supports `Callable` which does let you throw an exception.

Alternatively, you can catch checked exceptions and rethrow using an unchecked exception, such as one of the `PicocliException` subclasses.

File: library/src/test/java/bio/terra/pfb/utils/CompareOutputUtils.java (hunk 15 above)
Comment (pshapiro4broad) [reply library/src/test/java/bio/terra/pfb/utils/CompareOutputUtils.java:57]: Using `JSONAssert` you don't need to use ObjectMapper, and you don't need your `PfbJsonCompareIgnoringOrder` class either. I ran this locally and it seems to work:
```
  public static void compareJavaPfbWithPyPfb(
      String fileName, PfbCommandType commandType, String filePath, FileExtension fileExtension)
      throws IOException {
    PfbReader reader = new PfbReader();
    String pythonOutput =
        Files.readString(
            Paths.get(getPyPfbOutputFilePath(fileName, commandType, filePath, fileExtension)));
    String avroFilePath = getAvroFilePath(fileName, filePath);
    String javaPfbOutput =
        switch (commandType) {
          case SHOW_SCHEMA -> reader.showSchema(avroFilePath);
          case SHOW_METADATA -> reader.showMetadata(avroFilePath);
          case SHOW_NODES -> reader.showNodes(avroFilePath);
          default -> throw new InvalidObjectException("Invalid command type");
        };

    switch (fileExtension) {
      case JSON -> JSONAssert.assertEquals(pythonOutput, javaPfbOutput, true);
      case TXT -> assertThat(pythonOutput, equalTo(javaPfbOutput));
      default -> throw new InvalidObjectException("Invalid file extension");
    }
  }

  public static void compareJSONLineByLine(
      String fileName, PfbCommandType commandType, String filePath) throws IOException {
    PfbReader reader = new PfbReader();

    String avroFilePath = getAvroFilePath(fileName, filePath);
    List<String> javaOutput = reader.show(avroFilePath);
    Path file = Paths.get(getPyPfbOutputFilePath(fileName, commandType, filePath, JSON));
    List<String> lines = Files.readAllLines(file);
    for (int i = 0; i < lines.size(); i++) {
      JSONAssert.assertEquals(lines.get(i), javaOutput.get(i), true);
    }
  }
```

I also made some minor changes here with respect to reading the lines of the file using `Files.readAllLines()` and removed the code that catches `IOException`.

File: library/src/main/java/bio/terra/pfb/PfbReader.java (hunk 9 above)
Comment (snf2ye) [reply library/src/main/java/bio/terra/pfb/PfbReader.java]: I went with the second option to catch the exception and rethrow as PicocliException.

File: cli/src/main/java/bio/terra/pfb/JavaPfbCommand.java (hunk 14 above)
Comment (snf2ye) [reply cli/src/main/java/bio/terra/pfb/JavaPfbCommand.java]: Yeah, I have had to constantly ignore warnings from sonarcloud about using system.out. 

Thinking back to the goal of this CLI is to more-or-less replicate the output from pyPFB. They only output the result of the command. This has been useful, because I've been able to directly write the output to JSON files/etc. 
So, I think I will go with your suggestion to go back to `System.out`.

Comment (pshapiro4broad) [reply cli/src/main/java/bio/terra/pfb/JavaPfbCommand.java]: I know you can customize the sonar warnings in the project dashboard (although I haven't done this and don't know exactly how it's done) to silence the `System.out` warning. Or you can change your code to use `PrintStream` and isolate the reference to `System.out` in a single place and then use`@SuppressWarnings` to disable the warning in that one place. For an example of this see https://github.com/DataBiosphere/terra-data-catalog/blob/main/common/src/main/java/bio/terra/catalog/common/RequestContextCopier.java#L48

File: library/src/test/java/bio/terra/pfb/utils/CompareOutputUtils.java (hunk 15 above)
Comment (snf2ye) [reply library/src/test/java/bio/terra/pfb/utils/CompareOutputUtils.java:55]: Ah, ok! I see that JSON assert provides good error messages when the json doesn't match. 
Thanks!

File: library/src/main/java/bio/terra/pfb/PfbReader.java [hunk 28]
```diff
@@ -0,0 +1,166 @@
...
+import java.util.stream.Collectors;
+import org.apache.avro.Schema;
+import org.apache.avro.file.DataFileReader;
+import org.apache.avro.file.DataFileStream;
+import org.apache.avro.generic.GenericDatumReader;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.io.DatumReader;
+import org.apache.avro.specific.SpecificDatumReader;
+
+public class PfbReader {
```
[the remark this answers is not in the document]
Comment (snf2ye) [reply library/src/main/java/bio/terra/pfb/PfbReader.java:20]: Oh, great! That makes testing the static methods with mocking a little harder, but everything else is more straightforward.

File: library/src/test/java/bio/terra/pfb/PfbReaderTest.java [hunk 29]
```diff
@@ -0,0 +1,93 @@
...
+    for (String fileName : editedListOfTestFiles) {
+      CompareOutputUtils.compareJavaPfbWithPyPfb(fileName, SHOW_SCHEMA, "", JSON);
+    }
+  }
+
+  @Test
+  void showNodesTest() throws IOException {
```
Comment (pshapiro4broad) [inline library/src/test/java/bio/terra/pfb/PfbReaderTest.java:30]: Sonar is going to flag this because it doesn't contain any asserts. If you rename `compareJavaPfbWithPyPfb()` to a name with `assert` like `assertJavaPfbIsPyPfb()` then the sonar issue should go away.

File: library/src/test/java/bio/terra/pfb/utils/CompareOutputUtils.java [hunk 30]
```diff
@@ -0,0 +1,103 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CompareOutputUtils {
+  private static final Logger logger = LoggerFactory.getLogger(CompareOutputUtils.class);
+
+  public static void compareJavaPfbWithPyPfb(
+      String fileName, PfbCommandType commandType, String filePath, FileExtension fileExtension)
+      throws IOException {
+    PfbReader reader = new PfbReader();
```
Comment (pshapiro4broad) [inline library/src/test/java/bio/terra/pfb/utils/CompareOutputUtils.java:24]: Now that all members of `PfbReader` are `static` you no longer need to create it here, and you can call it below using `PfbReader.showSchema(...)`

**Review by pshapiro4broad — APPROVED** [review]
Looks OK, a few more comments on the new code

File: cli/src/main/java/bio/terra/pfb/JavaPfbCommand.java [hunk 31]
```diff
@@ -1,15 +1,51 @@
...
+import static picocli.CommandLine.*;
 
+import bio.terra.pfb.models.PfbCommand;
+import bio.terra.pfb.models.PfbCommandOption;
+import bio.terra.pfb.models.commands.*;
+import java.io.IOException;
 import picocli.CommandLine;
 
+// Sonar complains about the use of System.out.println, but it seems right the right logging choice
+// for a CLI
```
Comment (calypsomatic) [inline cli/src/main/java/bio/terra/pfb/JavaPfbCommand.java:13]: Nit: extra "right"

**Review by calypsomatic — APPROVED** [review]

File: library/build.gradle [hunk 32]
```diff
@@ -19,9 +20,42 @@ repositories {
...
+    implementation 'org.xerial.snappy:snappy-java:1.1.10.3'
+    testImplementation 'org.json:json:20230618'
+    testImplementation 'com.fasterxml.jackson.core:jackson-databind:2.15.2'
+    testImplementation 'org.skyscreamer:jsonassert:1.2.3'
+}
+
+sourceSets.main.java.srcDirs += ["${buildDir}/generated-src"]
+spotless {
+    java{
+        targetExclude "${buildDir}/**", "build/generated-src/**"
```
[the remark this answers is not in the document]
Comment (snf2ye) [reply library/build.gradle:35]: I don't understand why, but the $buildDir reference doesn't work in GHA
